gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/watermarker/PartitionLevelWatermarker.java

```diff
@@ -0,0 +1,339 @@
+      }
+    }
+
+    log.info("Loaded watermarks from previous state " + this.previousWatermarks);
+
+    this.expectedHighWatermarks.putAll(this.previousWatermarks);
+    this.maxPartitionsPerDataset =
+        sourceState.getPropAsInt(MAX_PARTITIONS_IN_WATERMARK_PER_DATASET_KEY,
+            DEFAULT_MAX_PARTITIONS_IN_WATERMARK_PER_DATASET);
+  }
+
+  /**
+   * Initializes the expected high watermarks for a {@link Table}
+   * {@inheritDoc}
+   * @see gobblin.data.management.conversion.hive.watermarker.HiveSourceWatermarker#onTableProcessBegin(org.apache.hadoop.hive.ql.metadata.Table, long)
+   */
+  @Override
+  public void onTableProcessBegin(Table table, long tableProcessTime) {
+
+    Preconditions.checkNotNull(table);
+
+    if (!this.expectedHighWatermarks.hasPartitionWatermarks(tableKey(table))) {
+      this.expectedHighWatermarks.setPartitionWatermarks(tableKey(table), Maps.<String, Long> newHashMap());
+    }
+  }
+
+  /**
+   * Adds an expected high watermark for this {@link Partition}. Also removes any watermarks for partitions being replaced.
+   * Replace partitions are read using partition parameter {@link AbstractAvroToOrcConverter#REPLACED_PARTITIONS_HIVE_METASTORE_KEY}
+   *
+   * {@inheritDoc}
+   * @see gobblin.data.management.conversion.hive.watermarker.HiveSourceWatermarker#onPartitionProcessBegin(org.apache.hadoop.hive.ql.metadata.Partition, long, long)
+   */
+  @Override
+  public void onPartitionProcessBegin(Partition partition, long partitionProcessTime, long partitionUpdateTime) {
+
+    Preconditions.checkNotNull(partition);
+    Preconditions.checkNotNull(partition.getTable());
+
+    if (!this.expectedHighWatermarks.hasPartitionWatermarks(tableKey(partition.getTable()))) {
+      throw new IllegalStateException(String.format(
+          "onPartitionProcessBegin called before onTableProcessBegin for table: %s, partitions: %s",
+          tableKey(partition.getTable()), partitionKey(partition)));
+    }
+
+    // Remove dropped partitions
+    Collection<String> droppedPartitions =
+        Collections2.transform(AbstractAvroToOrcConverter.getDropPartitionsDDLInfo(partition),
+            new Function<Map<String, String>, String>() {
+              @Override
+              public String apply(Map<String, String> input) {
+                return PARTITION_VALUES_JOINER.join(input.values());
+              }
+            });
+
+    this.expectedHighWatermarks.removePartitionWatermarks(tableKey(partition.getTable()), droppedPartitions);
+    this.expectedHighWatermarks.addPartitionWatermark(tableKey(partition.getTable()), partitionKey(partition),
+        partitionUpdateTime);
+  }
+
+  /**
+   * Always 0. As this is a {@link PartitionLevelWatermarker}
+   *
+   * {@inheritDoc}
+   * @see gobblin.data.management.conversion.hive.watermarker.HiveSourceWatermarker#getPreviousHighWatermark(org.apache.hadoop.hive.ql.metadata.Table)
+   */
+  @Override
+  public LongWatermark getPreviousHighWatermark(Table table) {
```

Comment:
I am a bit confused here. Are callers supposed to know not to call getPreviousHighWatermark(table) if we are using PartitionLevelWatermarker implementation ? This looks like a break in abstraction.